include/zephyr/pm/pm.h

```diff
@@ -97,6 +97,33 @@ struct pm_notifier {
  */
 #define PM_NOTIFIERS_STATES(node_id) &Z_PM_NOTIFIERS_NAME(node_id)
 
+#ifdef CONFIG_PM
+/**
+ * @brief Application/driver helper for defining an instance-specific
+ *	  pm_notifier struct.
+ *
+ * @param index Application/driver instance index
+ * @param state_direction pm_state_direction: state entry, exit, or both
+ * @param callback_fn callback function
+ */
+#define PM_NOTIFIER_INST_DEFINE(index, state_direction, callback_fn)	\
+static struct pm_notifier pm_notifier_##index = {			\
+	.direction = state_direction,					\
+	.callback = callback_fn,					\
+	.ctx = DEVICE_DT_INST_GET(index),				\
+}
+
+/**
+ * @brief Retrieve the name of a pm_notifier struct instance.
+ *
+ * @param index Application/driver instance index
+ */
+#define PM_NOTIFIER_INST(index) pm_notifier_##index
```

Comment:
ditto for naming (`_DT_`).

What about drivers not using instance indices? `DT_INST` macros are not always used by drivers. I'd suggest having a non-inst version (ie taking node_id) and a inst version, expanding to the right node_id from an instance. You can use the unique DT_ORD number internally for naming purposes.

Reply:
Note that these helper macros aren't directly devicetree dependent, as they also need to work from application code. Index in this case can be anything, A device instance index, ORD number, or any other string/number.

Should I replace the `_INST_` part of the name to remove the confusion?

Reply:
I see I forgot to push my change that removes the use of `DEVICE_DT_INST_GET` from said macro. Above comment will apply after that.

Reply:
> Note that these helper macros aren't directly devicetree dependent, as they also need to work from application code. Index in this case can be anything, A device instance index, ORD number, or any other string/number.
> 
> Should I replace the `_INST_` part of the name to remove the confusion?

Then we must be careful, because we can have potential collisions. I'd have a generic API, and a device specific variant built on top, that uses ordinals as its internal naming scheme. The generic should document avoidance of numeric only names.

Reply:
This should be generic, we shouldn't assume that the context is always a device, make it a parameter and probably replace index with some other identifier (like _name).

Reply:
Added generic function and DT node/instance variations that make use of it.

Let me know if you find the implementation acceptable.